pandas/core/groupby/generic.py

```diff
@@ -256,16 +263,25 @@ def aggregate(self, func=None, *args, engine=None, engine_kwargs=None, **kwargs)
             if self.grouper.nkeys > 1:
                 return self._python_agg_general(func, *args, **kwargs)
 
-            try:
-                return self._python_agg_general(func, *args, **kwargs)
-            except (ValueError, KeyError):
-                # TODO: KeyError is raised in _python_agg_general,
-                #  see see test_groupby.test_basic
-                result = self._aggregate_named(func, *args, **kwargs)
+            if isinstance(
+                self._selected_obj.index, (DatetimeIndex, TimedeltaIndex, PeriodIndex)
```

Comment:
this is the one place where i'm _not_ using `self._selected_obj._can_use_libreduction`, as doing so would require 2.5 more kludges to get the tests passing:


1) below on 283 after ret = create_series_with_explicit_dtype would need to do

```
            # Inference in the Series constructor may not infer
            #  custom EA dtypes, so try here
            ret = maybe_cast_result(ret._values, obj, numeric_only=True)
            ret = Series(ret, index=index, name=obj.name)
```

1.5) in create_series_with_explicit_dtype would need to change dtype_if_empty=object to dtype_if_empty=obj.dtype (which im not 100% sure  about)

2) The kludge [here](https://github.com/pandas-dev/pandas/pull/34997/files?file-filters%5B%5D=.py#diff-bfee1ba9e7cb79839776fac1a57ed940R507) would have to be amended from `and name in output.index` to `and (name in output.index or 0 in output.index)`